Prometheus dashboard add-on kube application ( #3118 )

jayunit100 opened the pull request:
Implementation of issue #3118.

Here's a first pass at promdash + prometheus integration into addons.

With this, you can pretty much just run both out of the box, it layers on top of the work done already by @mwielgus.  

See the README for directions, mostly its just (1) run a docker command locally and then (2) create the pod.  

I think we can make this into a replication controller as a next step.

cc @timothysc @rrati

mwielgus committed:
Experimental config for Prometheus

jayunit100 committed:
Promdash integration, minor changes to marcin's prometheus pods and v1beta3 implementation, and cold-start specification for prometheus+promdash deployment, and move to contrib.

jayunit100 commented:
For those new to prometheus and also for the sake of verification, I've attached a screenshot of a working promdash/prometheus pair which spins up from following the README directions... 

![screenshot from 2015-05-06 23 50 47](https://cloud.githubusercontent.com/assets/826111/7508223/c710dec4-f44a-11e4-8ae4-20404bb39712.png)

wojtek-t commented:
cc @fgrzadkowski FYI

mbforbes commented:
shot in the dark, but assigning to @zmerlynn for addons?

jayunit100 commented:
Hi.  I updated the boilerplate checker, so that it only verifies license header in `.sh` files.  I put that as a *separate* commit.

eparis commented:
I'd rather see the boilerplate change as a separate PR.  I'm still trying to decide how I feel about it...

timothysc commented on cluster/addons/prometheus/prometheus.yaml:
```diff
@@ -0,0 +1,41 @@
+apiVersion: v1beta1
```
why not beta3 for *?

timothysc commented on cluster/addons/prometheus/prometheus.yaml:
```diff
@@ -0,0 +1,41 @@
+            hostPort: 3000
+        env:
+          - name: "DATABASE_URL"
+            value: "sqlite3:/promdash/file.sqlite3"
```
Can we comment out a block for folks to enable volume mounts for persistence. 

/cc @markturansky

eparis commented on cluster/addons/prometheus/run_prometheus.sh:
```diff
@@ -0,0 +1,103 @@
+
+  ### Allow prometheus to scrape multiple targets
+  ### i.e. -t KUBERNETES_RO,MY_METRICS_ABC
+  for target in ${2//,/ }; do
```
Is this a bash-ism, or it generic enough that leaving /bin/sh at the top is ok?

timothysc commented on cluster/addons/prometheus/README.md:
```diff
@@ -0,0 +1,70 @@
+
+- We should publish this image into the kube/ namespace.
+- Possibly use postgre or mysql as a promdash database.
+- Setup high availability via NFS
```
push gateway (https://github.com/prometheus/pushgateway)?

timothysc commented:
@vishh - eventually I could see this as a roll up with heapster.

vishh commented:
@timothysc: Ack. I plan to start working on prometheus support in heapster soon. I just hope we don't re-invent heapster and end up maintaining one more component.

timothysc commented:
@vishh +1

jayunit100 commented:
I'll implement these changes and update to bash and retest........ then will update pr

jayunit100 commented:
okay ! Implemented the v1beta3 , tested, did some minor cleanups, rebased, removed my boilerplate changes, and layered commits properly.  Ready for final  review !

jayunit100 commented:
hi, @zmerlynn @timothysc looks good to merge now ?  would like to run this as a pod in our cluster starting monday.

timothysc commented:
@fgrzadkowski @zmerlynn @quinton-hoole ping ^

quinton-hoole commented:
Sorry, we're all a bit swamped with v1.0 blockers at the moment - will take me a few days to get to this.

bprashanth commented on cluster/addons/prometheus/README.md:
```diff
@@ -0,0 +1,71 @@
+http://service_address:service_port/metrics.
+
+# Purpose
+The purpose of the setup is to gather performance-related metrics during load
```
Its a little confusing that this is a cluster addon, for this to be useful doesn't it have to be a standalone thing that multiple load tests can send their metrics to across a week? how do other e2e clusters discover the dashboard?

jayunit100 commented:
the metrics are already made available by the api server and they are in prometheus format ... prom dash and prometheus just scrape them from the kubkubernetes_ro endpoints

jayunit100 commented:
Clarification: prometheus scrapes... prom dash gives a graphing ui

bprashanth commented:
I guess im asking if the dashboard is brouht up/down with each test run? (haven't looked at your code yet, but sounds like that must be the case unless you're always scraping a known apiserver endpoint that's the same for all test clusters).

jayunit100 commented:
good question. so the way it works is that the kubernetes_ro endpoint is always available to any container... and that is the magic that allows the prometheus server to launch from a pod.  Makes sense?

bprashanth commented:
That I know :)
'pologies, when i said test run i meant test suite. I would find a dashboard that all my e2e test suites could send, eg: latency of a list/watch to very useful (or aggregate duration of all the lists from all the tests for a whole week).

jayunit100 commented:
This Sounds like a specific question about a generic feature add on ... i was hoping folks would evolve this as a base starting point : there are many ways to leverage prometheus

timothysc commented:
@bprashanth We currently use Prometheus as an external component to our cluster, b/c we tare-down and rebuild our cluster nightly.  By doing this, we preserve history.   

The add-on PR is meant to be a simple introspective service which "could" store it's data across runs by mounting a network volume.

zmerlynn commented:
This feels more like a `contrib/` PR than

jayunit100 commented:
Based on experience with other projects;  Contrib/ implies it's a third party artifact which isn't necessarily tied to the core project.  Is that the case in kube?

If yes...

Prometheus is, pretty deeply integrated with kubernetes as of today... and so this is just the next logical step in supporting prometheus.

If no....

Then I guess it might belong in contrib/... but maybe then the use of contrib is for a different purpose then I've guessed above.

Either way is fine with me but I'd prefer to iterate tThis code in a place which implies it is a  supported compoment.

zmerlynn commented:
Right now, everything in the `addons` directory is a pod of some sort that gets loaded by Salt through one means or another. This doesn't seem to fit that that pattern. Hm. I'm not sure where it belongs.

timothysc commented:
@vishh Given that you would be the heapster integration person, do you have opinions on the matter?

zmerlynn commented:
Note that what I said is kind of an artifact of just what's already been built. This is definitely an addon, but everything else in that dir so far is something we have an option of launching at cluster creation. I suppose this isn't too far off the first steps of Elasticsearch, though.

vishh commented:
IIUC, Kubernetes is not deeply tied to Prometheus. The fact that we use
prometheus client libraries does not mean we are tied to Prometheus. We
could very well move to a different metrics client library in the future.
This PR feels like a Contrib/ to me as well. We are trying to bring down
the number of external dependencies in the project and another addon is not
ideal at this stage. The core system can function without a monitoring
backend. But the existing addons like kube2sky and heapster are required
for the core to function.

Once heapster scrapes the component metrics, we can use other monitoring
solutions too for this purpose.

On Thu, May 14, 2015 at 8:06 AM, Zach Loafman <notifications@github.com>
wrote:

> Note that what I said is kind of an artifact of just what's already been
> built. This is definitely an addon, but everything else in that dir so far
> is something we have an option of launching at cluster creation. I suppose
> this isn't too far off the first steps of Elasticsearch, though.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/7878#issuecomment-102065889>
> .
>

jayunit100 commented:
ok.  well if the consensus to move it to `contrib/` i can do that.  anyone else want to vote ? looks like 2 in favor of contrib, 1 in favor of addon.

jayunit100 commented:
ok, moved to contrib & rebased :)

timothysc commented:
thx jay! 

@zmerlynn any other comments?

keegancsmith commented:
This was merged in a while ago, but not sure of a better place to bring this up. Prometheus now has built in Kubernetes discovery and the example config will automatically scrape all the `kube-system` services + services which have the annotation `"prometheus.io/scrape": "true"`. https://github.com/prometheus/prometheus/blob/master/documentation/examples/prometheus-kubernetes.yml

jayunit100 commented:
Hi keegan thanks for bringing this up.  We have prometheus and promdash
integrated in the contrib/prometheus stuff. So long as the functionality is
retained, I think replacing parts with the idioms that re being used in
github.com/prometheus/... makes sense.  As I recall, the prometheus docker
images were out of data and promdash was a 3rd party extension at the time
that we created this.




On Thu, Oct 1, 2015 at 5:36 AM, Keegan Carruthers-Smith <
notifications@github.com> wrote:

> This was merged in a while ago, but not sure of a better place to bring
> this up. Prometheus now has built in Kubernetes discovery and the example
> config will automatically scrape all the kube-system services + services
> which have the annotation "prometheus.io/scrape": "true".
> https://github.com/prometheus/prometheus/blob/master/documentation/examples/prometheus-kubernetes.yml
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/7878#issuecomment-144670656>
> .
>



-- 
jay vyas

fgrzadkowski commented:
FYI @mwielgus @jszczepkowski @piosz
